### ISSUE TITLE
Tobii mouse tracking is very jumpy

### ISSUE THREAD

**Issue by kmcnaught** [OptiKey/OptiKey#671]
I've always found mouse control with the Tobii EyeX much much jumpier than when I use the same eye tracker with the built in Windows Eye Control, or Tobii Gaze Point. I thought it was a function of Optikey's own fixation-processing for mouse control, but I've investigated a bit further and it looks like the way we're hooking up to the tobii data stream isn't correct. 

You can see the kind of jumpiness I'm referring to here (video for bug #670):
https://www.youtube.com/watch?v=ysc6Yw6IoSU&feature=youtu.be 

In `TobiiEyeXPointService.cs`, we set up the stream according to the setting for `TobiiEyeXProcessingLevel` which can be `[None, Low, Medium, High]`. The default is `Medium`. As a user I would expect High to mean a smoother tracking, at the cost of slower reaction. However, what actually happens is that a processing level of `Medium` or `High` means that Optikey subscribes to a `FixationDataStream` with the EyeX engine rather than a `GazePointDataStream`. This means that `Medium` actually isn't a smoothly varying gaze stream, but a series of fixations jumping around. We get away with it while selecting keys in Optikey, but when controlling a cursor this makes the cursor really jumpy and plays havoc with Optikey's own fixation algorithms. Although the video isn't too bad, I often find I can't actually get a mouse fixation to trigger at all in the location I want. 

Code is here for reference:
https://github.com/OptiKey/OptiKey/blob/073d19e9850510e0b20c85ec03826ff879c94bbc/src/JuliusSweetland.OptiKey.Core/Services/TobiiEyeXPointService.cs#L82-L128

Current logic is:
ProcessingLevel == None ->  use GazePointDataStream, Unfiltered 
ProcessingLevel == Low ->  use GazePointDataStream, LightlyFiltered
ProcessingLevel == Medium ->  use FixationDataStream , Sensitive 
ProcessingLevel == High ->  use FixationDataStream , Slow

I'd propose instead:
ProcessingLevel == None ->  use GazePointDataStream, Unfiltered 
ProcessingLevel == Low ->  use GazePointDataStream, LightlyFiltered
ProcessingLevel == Medium ->  use GazePointDataStream, LightlyFiltered + Kalman Filter, with relatively high confidence in new data (i.e. less smoothing than current implementation)
ProcessingLevel == High ->  use GazePointDataStream, LightlyFiltered + Kalman Filter as it currently is (quite a lot of smoothing)

**Comment by JuliusSweetland** [comment]
@kmcnaught What you're saying makes sense, as does your proposal. I don't understand the difference between the proposed Medium and High processing levels though? Would it not make more sense to offer new settings:

Stream Processing: 
Off (unfiltered)
Smooth (lightly filtered)
Smooth+Kalman (lightly filtered+kalman)
Fast Fixations (sensitive fixation stream)
Fast Fixations+Kalman (sensitive fixation stream + kalman)
Slow Fixations (slow fixation stream)
Slow Fixations+Kalman (slow fixation stream + kalman)

That still exposes all current functionality, but let me know whether you think this would make sense to users.

We would move the default to Smooth, or Fast Fixations.

We would need to provide an auto migration for current values:

None -> Off
Low -> Smooth
Medium -> Fast Fixations
High -> Slow Fixations

Thoughts?

**Comment by kmcnaught** [comment]
I'm not sure when it would ever be appropriate to use a fixation stream instead of a gaze stream? Don't we always want "current gaze point" rather than "current fixation" when consuming a stream of gaze data? The fixation stream will effectively be discretised jumps of gaze data, not a continuously varying stream.

The only time we're interested in fixations is at the point at which a user is performing a mouse action "at position", at which point you presumably apply some fixation logic in Optikey itself. It doesn't make sense to me to ever have the 'raw' data input as a stream of fixations, but it does make sense to have some processing which smooths / throws away outliers / whatever else the tobii engine does.

In terms of the last two suggestions - I tried the Kalman filter, and it works really nicely for smoothing things out, but it's pretty heavy-handed, so you get quite a sluggish response. The end result is about what I'd expect for "High Processing". The Kalman filter could be tuned so that it puts more weight (aka confidence) on new data vs existing predictions, which would make it less sluggish. We could default to having two settings for Kalman filtering parameters, where "High" processing gives the current Kalman filter behaviour, and "Medium" processing is akin to "Kalman Filtering Lite".

**Comment by JuliusSweetland** [comment]
@kmcnaught It's a tricky one, because as you say, fixations work well when selecting keys on the keyboard. It's only when the user requires smooth movement (mouse control) where it is a problem. That's my argument for retaining some fixation processing on the incoming stream, so that we have the ability to default to a fixation stream for key only versions of optikey (symbol and chat).

**Comment by JuliusSweetland** [comment]
@kmcnaught Forgot to add: "thoughts?" :)

**Comment by kmcnaught** [comment]
I guess I'm not convinced that fixation processing is even helpful for selecting keys on the keyboard. Having said that, I've just spent some time comparing different levels of processing (with the current implementation of Low, Med, High) and it's really hard to make an objective assessment. I'm still having (separate) trouble with Optikey's fixation detection for mouse events - it's easy to dwell on the target if your initial fixation lands in exactly the right place, but if you need to adjust at all you end up snapped to the original fixation, or having to jump away and back in again until you get lucky. So I think perhaps I'm experiencing the conflation of two things, both of which are particularly pertinent when using my EyeX. 

I guess I should restate my original issue more broadly: The EyeX settings should be more forgiving out of the box (not requiring a user to turn on a bunch of extra processing before they can do any mouse control). Currently it's significantly jumpier than what you get from the Gaze Point app and I feel this makes Optikey look bad. 

Related to this, two concrete things that could be improved:
1) I'd like Optikey's fixation logic for mouse clicks to be more forgiving
2) I'd like the Kalman filter to be available in much milder form.

Point (2) can be somewhat solved by exposing the KF parameters to the settings UI, though I'd suggest it's probably best to be available to a user in predefined levels (e.g. low/med/high) rather than raw params. And perhaps it's appropriate to default to having _some_ kalman filtering on for the Tobii Gaming trackers, until it looks approximately like the Gaze Point app. 

Point (1) is something I've often thought about but never got round to investigating.

**Comment by JuliusSweetland** [comment]
I think the issue with mouse fixations is that maybe the initial lock in
area is too large, which means you cannot easily adjust it. There has been
talk of changing the approach so that the user flies into the point, making
adjustments as they go. I think Adam said he might look at this.

The rest of your points I agree with.

**Comment by AdamRoden** [comment]
@JuliusSweetland @kmcnaught I've been vastly upgrading my Optikey experience with a set of seamless dynamic keyboards, specialized dynamic keys, and optimized Management Console settings. Here are some of my observations:

- Upgrading from the Tobii 4c to the Tobii Eye Tracker 5 did not result in any noticeable differences related to Optikey.
- Turning the Kalman filter off is the single most significant performance improvement that improves the eye gaze experience.
- I don't notice much difference in low/medium/high processing. Without reading your earlier comments, I had settled on medium in my settings.
- The Kalman filter is hugely detrimental to key selection, but may have a place in point selection or the magnetic cursor. 
- Personally, I have no issue with jitteriness because I have little use for the magnetic cursor and my mouse click speeds are so fast without it that it doesn't matter.

In conclusion, the default setting for the Kalman filter should be off, and if we are going to provide it as an option then we should allow it to be applicable under situational conditions, (i.e. 2nd step of mouse click or magnetic cursor.)

**Comment by kmcnaught** [comment]
@AdamRoden that's helpful feedback. Combining your thoughts and mine, I'm thinking that perhaps a reasonable option would be for the user-facing Kalman filter settings to be:

```
- Kalman filter ON/OFF 
    - KF processing level [Low/Medium/High]  (this would change the KF weights to 3 pre-determined levels)
    - Use for point selection ON/OFF  (default on, when KF switched on)
    - Use for magnetic cursor ON/OFF  (default on/off, not sure)
    - Use for key selection  (default off) 
```

Depending how this felt, I'd be tempted to have the Kalman filter on for the EyeX by default, but only affecting point selection, and with the lowest processing level (much lower than the current defaults, which would probably become "high", or somewhere between "medium" and "high"). The current defaults for point selection (dwell timings) are quite high (slow), which currently makes it very hard to use point selection accurately with my 4C with default settings.

**Comment by AdamRoden** [comment]
@kmcnaught I find that when my cursor is too jumpy it is usually resolved by recalibrating.  Here are my proposed settings:

>Key Selection
> -Smoothing Level [**Off**/Low/Medium/High]
>
>Magnetic Cursor
> -Smoothing Level [Off/Low/**Medium**/High]
>
>Point Selection (Area to Magnify)
> -Smoothing Level [**Off**/Low/Medium/High]
> -Cursor Size [**None**/Small/Medium/Large]
> -Lock On Time [300]
> -Completion Time [0]
> -Radius [300]
> -% of Screen to Magnify [10]
>
>Point Selection (When Magnified)
> -Smoothing Level [Off/Low/**Medium**/High]
> -Cursor Size [None/Small/**Medium**/Large]
> -Lock On Time [100]
> -Completion Time [300]
> -Radius [60]
> -% of Screen to Fill [50]

**Comment by kmcnaught** [comment]
The point remains that out of the box, with a fresh calibration, Optikey's mouse control is noticeably jumpier than Tobii's Gaze Point app, and I think we should strive to give a comparable experience, or better. I was testing yesterday with a new calibration, but I don't always, because it's important for the UX to fail gracefully when calibration is a bit off. I'm also conscious that some users just have a worse calibration regardless, due to other factors, not always easily explained.

It feels a bit like we're trying add lots of extra logic to work around the fact that the Kalman Filter isn't actually doing what we want it to do. Looking more closely at the implementation, it isn't doing anything more than a moving average (i.e. always smoothing by the same amount regardless of the data coming in).

If you know a bit about Kalman Filters, this is because the implicit process model in Optikey's KF implementation is "your eyes haven't moved" + some uncertainty (they may have moved a bit locally, but probably not very far). The uncertainty is fixed, and given by `ProcessNoise` in the Settings. Because this uncertainty is constant, the measurement uncertainty is constant, and the prediction and update steps are rolled into one (we update everything for each new datapoint and don't take time into account), the uncertainty of our position estimate quickly converges to a constant, where it remains forever more. This is a bit counterintuitive because you might expect that the uncertainty should shrink over time while you're fixating, and then grow when your eyes jump somewhere new. But the process model doesn't know about saccades so this doesn't happen.

TL;DR: the main problem is that the process model is assuming you are fixating, and when you are, the smoothing is helpful. But once your eyes jump to a new location, this process model is completely wrong and the filtering is unhelpful.

I spent some time yesterday exploring a couple of possible solutions, which let us throw away our rolling estimates once we believe a saccade has occurred. This gives a nice balance between smoothness while fixating, and rapid response when saccading. I'll have to get back later with some data / videos because I have other work to do! I got a bit nerd-sniped by this one :)

**Comment by kmcnaught** [comment]
PS - definitely agree on re-branding the kalman filter as "smoothing" at the user-facing stage.

**Comment by JuliusSweetland** [comment]
Looking forward to seeing the data when you have time Kirsty.

**Comment by AdamRoden** [comment]
@kmcnaught I agree with your analysis. 
I will clarify my current opinions: 
For key selection the fastest response will always come when the KF is off. 
However, anytime we are showing a cursor, or in the future a potential gaze tracing bubble, the best experience will come with the KF on at some level. 
Now the grey area is during a 2-step mouse click. I think the fastest method is to wait to show the cursor and utilize the KF until an area of the screen is magnified. However, we should offer the option of showing the cursor during both steps.

P.S. KF turned off, and other settings fine-tuned, my mouse selection times are faster than I was able to achieve with my previous favorite, "continuous zoom click" by Tobii, so I no longer desire the feature.

**Comment by siamakmu** [comment]
Hi dear Julius,
Is there a difference between Tobii C4 and Tobii EyeX in terms of performance in optikey software? How about Tobii 5?

**Comment by JuliusSweetland** [comment]
Hi - yes the EyeX is the oldest of the 3. The 4C is very good, but not
easily available for purchase. The Eye Tracker 5 is the latest iteration of
the tracker and the only one (as far as I know) that is now available to
buy. It is good. I actually prefer the 4C, but they are pretty similar.

**Comment by alexandre-mbm** [comment]
> It is good. I actually prefer the 4C, but they are pretty similar.

@JuliusSweetland, why?

**Comment by JuliusSweetland** [comment]
I was having some issues with the ET5, but I think I've figured out the problem. I think the ET5 is very good now

**Comment by alexandre-mbm** [comment]
[Tom's Hardware about the ET5](https://www.tomshardware.com/amp/news/tobii-eye-tracker-5-pc-games)
